src/BitIntegers.jl

```diff
@@ -453,6 +453,10 @@ function read(from::GenericIOBuffer, ::Type{T}) where {T<:XBI}
     return x
 end
 
+function read(io::IO, ::Type{T}) where {T<:XBI}
+    return first(reinterpret(T, read(io, sizeof(T))))
```

Comment:
Maybe this is preferable? Doesn't seem to impact performance either way:
```suggestion
    return reinterpret(T, read(io, sizeof(T)))[]
```

Reply:
I don't really mind. I just had a look at how it's done in Base, they have `read!(s, Ref{T}(0))[]::T` which seems to work here too. cf. https://github.com/JuliaLang/julia/blob/2479691a5a91051d0b71552ecbcf6724a45409a7/base/io.jl#L726

Reply:
Looks like using `read!` is the clear winner:
```julia
julia> using BitIntegers, BenchmarkTools

julia> function _read1(io::IO, ::Type{T}) where {T<:BitIntegers.XBI}
           first(reinterpret(T, read(io, sizeof(T))))
       end
_read1 (generic function with 1 method)

julia> function _read2(io::IO, ::Type{T}) where {T<:BitIntegers.XBI}
           reinterpret(T, read(io, sizeof(T)))[]
       end
_read2 (generic function with 1 method)

julia> function _read3(io::IO, ::Type{T}) where {T<:BitIntegers.XBI}
           read!(io, Ref{T}(0))[]::T
       end
_read3 (generic function with 1 method)

julia> x = [0x01; fill(0x00, 63)];

julia> @benchmark _read1(IOBuffer(x), UInt512)
BenchmarkTools.Trial:
  memory estimate:  304 bytes
  allocs estimate:  4
  --------------
  minimum time:     363.678 ns (0.00% GC)
  median time:      382.541 ns (0.00% GC)
  mean time:        458.851 ns (3.47% GC)
  maximum time:     14.768 μs (95.90% GC)
  --------------
  samples:          10000
  evals/sample:     208

julia> @benchmark _read2(IOBuffer(x), UInt512)
BenchmarkTools.Trial:
  memory estimate:  304 bytes
  allocs estimate:  4
  --------------
  minimum time:     393.292 ns (0.00% GC)
  median time:      405.394 ns (0.00% GC)
  mean time:        486.287 ns (3.41% GC)
  maximum time:     16.297 μs (96.94% GC)
  --------------
  samples:          10000
  evals/sample:     202

julia> @benchmark _read3(IOBuffer(x), UInt512)
BenchmarkTools.Trial:
  memory estimate:  224 bytes
  allocs estimate:  3
  --------------
  minimum time:     128.423 ns (0.00% GC)
  median time:      131.847 ns (0.00% GC)
  mean time:        164.683 ns (5.26% GC)
  maximum time:     3.329 μs (92.45% GC)
  --------------
  samples:          10000
  evals/sample:     893
```

Reply:
Do you still want to keep the `read(from::GenericIOBuffer, ::Type{T})` method?
```julia
julia> @benchmark read(IOBuffer(x), UInt512)
BenchmarkTools.Trial:
  memory estimate:  144 bytes
  allocs estimate:  2
  --------------
  minimum time:     157.345 ns (0.00% GC)
  median time:      159.992 ns (0.00% GC)
  mean time:        197.522 ns (3.44% GC)
  maximum time:     6.843 μs (96.45% GC)
  --------------
  samples:          10000
  evals/sample:     811
```

Reply:
> Do you still want to keep the read(from::GenericIOBuffer, ::Type{T}) method?

Yes! Try interpolating `x` in your benchmarks, the timing differences are big ;-)
Can you repost updated benchmarks for `_read1`, `_read2`, `_read3` ?

Reply:
Good point. I skipped interpolating here as we need to rewind the buffer for each sample:
```julia
julia> @benchmark _read1(seekstart($(IOBuffer(x))), $UInt512)
BenchmarkTools.Trial:
  memory estimate:  160 bytes
  allocs estimate:  2
  --------------
  minimum time:     251.242 ns (0.00% GC)
  median time:      255.153 ns (0.00% GC)
  mean time:        300.596 ns (2.28% GC)
  maximum time:     7.451 μs (94.64% GC)
  --------------
  samples:          10000
  evals/sample:     372

julia> @benchmark _read2(seekstart($(IOBuffer(x))), $UInt512)
BenchmarkTools.Trial:
  memory estimate:  160 bytes
  allocs estimate:  2
  --------------
  minimum time:     274.625 ns (0.00% GC)
  median time:      283.845 ns (0.00% GC)
  mean time:        336.276 ns (2.31% GC)
  maximum time:     9.395 μs (94.41% GC)
  --------------
  samples:          10000
  evals/sample:     296

julia> @benchmark _read3(seekstart($(IOBuffer(x))), $UInt512)
BenchmarkTools.Trial:
  memory estimate:  80 bytes
  allocs estimate:  1
  --------------
  minimum time:     14.192 ns (0.00% GC)
  median time:      16.081 ns (0.00% GC)
  mean time:        20.865 ns (7.90% GC)
  maximum time:     1.264 μs (97.48% GC)
  --------------
  samples:          10000
  evals/sample:     998

julia> @benchmark read(seekstart($(IOBuffer(x))), $UInt512)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.031 ns (0.00% GC)
  median time:      4.077 ns (0.00% GC)
  mean time:        4.652 ns (0.00% GC)
  maximum time:     124.857 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
```
The `read(from::GenericIOBuffer, ::Type{T})` does have a slight performance advantage still.

Reply:
I get very different times for `_read1` and `_read2`, e.g.  
```julia
julia> @btime _read1(seekstart($(IOBuffer(x))), UInt512);
  80.136 ns (2 allocations: 160 bytes)

julia> @btime _read2(seekstart($(IOBuffer(x))), UInt512);
  80.607 ns (2 allocations: 160 bytes)

julia> @btime _read3(seekstart($(IOBuffer(x))), UInt512);
  19.946 ns (1 allocation: 80 bytes)

julia> @btime read(seekstart($(IOBuffer(x))), UInt512);
  2.994 ns (0 allocations: 0 bytes)
```
Anyway, yes let's keep `read(from::GenericIOBuffer, ::Type{T})` for now (and I guess this was copied from `Base`, so if the performance would justify changing the implementation, it would probably apply to `Base` types too).